orttraining/orttraining/training_api/checkpoint.h

```diff
@@ -0,0 +1,156 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/platform/path_lib.h"
+#include "core/platform/env.h"
+#include "onnx/defs/tensor_proto_util.h"
+#include "orttraining/core/framework/checkpointing.h"
+#include "orttraining/training_api/interfaces.h"
+
+namespace onnxruntime {
+namespace training {
+
+// TODO: Rename to api after all major classes implemented.
+namespace api_test {
+
+template <typename T>
+struct TypedCheckpointProperty;
+
+/**
+ * @brief Base class for user defined checkpoint property.
+ */
+struct CheckpointProperty {
```

Comment:
In the interest of simplicity can we just have CheckpointProperty? do we need base and derived classes here? Not sure I understand what is the purpose of this.

Reply:
This is to support different data types, we could manage int/float/string typed property data in a vector.

Reply:
Also moved CheckpointProperty and TypedCheckpointProperty into orttraining/orttraining/training_api/checkpoint_property.h.

Reply:
Introduced PropertyBag to manage custom properties adding/serializing to tensorproto/retrieving. Easier for user to add property.